### PR TITLE
[flang] Assume matching shapes in elemental assignment with non-realloc lhs.

### PR DESCRIPTION
**Pull request by vzakhari**
The optimized bufferization pass cannot optimize very simple cases of elemental
assignments, because of the suboptimal checks order. This patch relies
on the fact that in a legal program the lhs and rhs of an assignment
have matching shapes, when lhs is not an allocatable and rhs is a result
of an elemental array operation.

### CODE REVIEW

File: flang/test/Integration/OpenMP/workshare-axpy.f90 [hunk 1]
```diff
@@ -13,7 +13,7 @@ subroutine sb1(a, x, y, z)
   integer :: a
   integer :: x(:)
   integer :: y(:)
-  integer :: z(:)
+  integer, allocatable :: z(:)
```
Comment (ivanradanov) [inline flang/test/Integration/OpenMP/workshare-axpy.f90:16]: Is this change required?

Comment (vzakhari) [reply flang/test/Integration/OpenMP/workshare-axpy.f90:16]: Yes, this change is required, if you want to keep testing the fact that the `fir.allocmem` and `fir.freemem` are placed in the `omp.single` regions. Those operations appear due to the optimized bufferization not being able to optimize the assignment without a runtime call, which requires the RHS to be computed into a temporary array.

With this patch the line 18 statement will be optimized such that there is no runtime call at all, and hence no temporary array allocation/deallocation.

Comment (ivanradanov) [reply flang/test/Integration/OpenMP/workshare-axpy.f90:16]: I see. thank you very much. I mistakenly thought the optimization was happening already.

File: flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp [hunk 2]
```diff
@@ -338,32 +338,33 @@ ElementalAssignBufferization::findMatch(hlfir::ElementalOp elemental) {
...
-    // the shapes do conform
-    if (match.assign.getRealloc())
+  // so we just have to check for the same ssa value.
+  //
+  // f2018 10.2.1.2 (3) requires the lhs and rhs of an assignment to be
+  // conformable unless the lhs is an allocatable array. In HLFIR we can
+  // see this from the presence or absence of the realloc attribute on
+  // hlfir.assign. If it is not a realloc assignment, we can trust that
+  // the shapes do conform.
+  if (match.assign.getRealloc()) {
```
Comment (jeanPerier) [inline flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:350]: Are there actually any cases were the check below will succeed?
Whole allocatable assignments are made to fir.ref<fir.box> LHS, and lower never generates fir.shape operands for the related declare/designate because the shape is dynamic.

We will need to optimize whole allocatable assignment, but this is its own can of worms (need to prove the LHS is allocated too, that it has the same length parameters for characters...).

Better to just return nullopt here I think and a TODO comment to optimize this case (maybe using code specialization?).

**Review by jeanPerier — APPROVED** [review]
Makes sense to me, thank you Slava!

File: flang/test/HLFIR/minloc-elemental.fir [hunk 3]
```diff
@@ -188,67 +188,65 @@ func.func @_QPtest_kind2_convert(%arg0: !fir.box<!fir.array<?xi32>> {fir.bindc_n
...
+// CHECK:             }
+// CHECK:             fir.result %[[VAL_26]] : i32
+// CHECK:           }
+// CHECK:           %[[VAL_37:.*]] = hlfir.as_expr %[[VAL_11]] move %[[VAL_3]] : (!fir.ref<!fir.array<1xi16>>, i1) -> !hlfir.expr<1xi16>
+// CHECK:           fir.do_loop %[[VAL_38:.*]] = %[[VAL_9]] to %[[VAL_9]] step %[[VAL_9]] unordered {
+// CHECK:             %[[VAL_39:.*]] = hlfir.apply %[[VAL_37]], %[[VAL_38]] : (!hlfir.expr<1xi16>, index) -> i16
+// CHECK:             %[[VAL_40:.*]] = fir.convert %[[VAL_39]] : (i16) -> i32
+// CHECK:             %[[VAL_41:.*]] = hlfir.designate %[[VAL_13]]#0 (%[[VAL_38]])  : (!fir.box<!fir.array<?xi32>>, index) -> !fir.ref<i32>
+// CHECK:             hlfir.assign %[[VAL_40]] to %[[VAL_41]] : i32, !fir.ref<i32>
+// CHECK:           }
```
Comment (tblah) [inline flang/test/HLFIR/minloc-elemental.fir:247]: In the future please could you not regenerate the entire test when only a small part has changed. This makes it hard work to find the change. AFAICT the only part that changed was this end part: with the elemental converted to a do-loop.

**Review by tblah — APPROVED** [review]
LGTM, thanks

File: flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp (hunk 2 above)
Comment (vzakhari) [reply flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:350]: You are right, Jean.  Those conditions are not very useful, because the LHS shape will always be null.  I will remove them. Thanks!

File: flang/test/HLFIR/minloc-elemental.fir (hunk 3 above)
Comment (vzakhari) [reply flang/test/HLFIR/minloc-elemental.fir:247]: I will try :)
While looking through these checks I found the issues with using the exact names like `%c2147483647_i32`, `%false`, etc., so I decided to make it "trully" auto-generated now with hopes that the future updates will be more contained (assuming the checks generator script will be used).

Thanks for the review, Tom!

**Review by jeanPerier — APPROVED** [review]
Thanks
